python/TestHarness/schedulers/Scheduler.py

```diff
@@ -0,0 +1,539 @@
+from time import sleep
+from collections import deque
+from MooseObject import MooseObject
+from TesterData import TesterData
+from signal import SIGTERM
+import os, sys, re, copy
+from contrib import dag
+
+import util
+from multiprocessing.pool import ThreadPool
+import threading # for thread locking
+import multiprocessing # for timeouts
+
+class SchedulerError(Exception):
+    pass
+
+class Scheduler(MooseObject):
+    """
+    Base class for handling how jobs are launched. To use this class, call .schedule()
+    and supply a list of testers to schedule.
+
+    Syntax:
+       .schedule([list of tester objects])
+
+    A list of testers will be added to a queue and begin calling the derived run method
+    immediately. You can continue to add more testers to the queue in this fashion.
+
+    Once you schedule all the testers you wish to test, call .waitFinish() to wait until
+    all tests have finished.
+
+    As tests finish individually, the scheduler will call back handleTestStatus(tester)
+    """
+
+    @staticmethod
+    def validParams():
+        params = MooseObject.validParams()
+        params.addRequiredParam('average_load',          64.0, "Average load to allow")
+        params.addRequiredParam('max_processes',         None, "Hard limit of maxium processes to use")
+
+        return params
+
+    def __init__(self, harness, params):
+        MooseObject.__init__(self, harness, params)
+
+        ## The test harness to run callbacks on
+        self.harness = harness
+
+        # Retrieve and store the TestHarness options for use in this object
+        self.options = harness.getOptions()
+
+        # The Scheduler class can be initialized with no "max_processes" argument and it'll default
+        # to a soft limit. If however a max_processes  is passed we'll treat it as a hard limit.
+        # The difference is whether or not we allow single jobs to exceed
+        # the number of slots.
+        if params['max_processes'] == None:
+            self.job_slots = 1
+            self.soft_limit = True
+        else:
+            self.job_slots = params['max_processes'] # hard limit
+            self.soft_limit = False
+
+        # Requested average load level to stay below
+        self.average_load = params['average_load']
+
+        # Initialize runner pool based on available slots
+        self.tester_pool = ThreadPool(processes=self.job_slots)
+
+        # Initialize status pool to only use 1 process (to prevent status messages from getting clobbered)
+        self.status_pool = ThreadPool(processes=1)
+
+        # Thread Locking
+        self.thread_lock = threading.RLock()
```

Comment:
I did some reading on this. "R" is for reentrant which means the thread that locks it can continue to pass through barriers protected with the same lock. My opinion is that unless this is something we need, we probably shouldn't use it because it might actually obscure logic problems with locking. Use "Lock"instead.